source/common/upstream/upstream_impl.cc

```diff
@@ -725,15 +731,26 @@ bool BaseDynamicClusterImpl::updateDynamicHostList(const HostVector& new_hosts,
             (*i)->healthFlagSet(Host::HealthFlag::FAILED_EDS_HEALTH);
             // If the host was previously healthy and we're now unhealthy, we need to
             // rebuild.
-            health_changed |= previously_healthy;
+            hosts_changed |= previously_healthy;
           } else {
             (*i)->healthFlagClear(Host::HealthFlag::FAILED_EDS_HEALTH);
             // If the host was previously unhealthy and now healthy, we need to
             // rebuild.
-            health_changed |= !previously_healthy && (*i)->healthy();
+            hosts_changed |= !previously_healthy && (*i)->healthy();
           }
         }
 
+        // Did metadata change?
+        const bool metadata_changed =
+            !Protobuf::util::MessageDifferencer::Equivalent(host->metadata(), (*i)->metadata());
+        if (metadata_changed) {
+          (*i)->metadata(host->metadata());
+          (*i)->canary(host->canary());
```

Comment:
Seems odd here to do the canary update inside the metadata check. Should this be an independent check for people that use canary but not metadata?

Reply:
@mattklein123 my thinking is that canary is encoded under `envoy.lb/canary`, so for it to change metadata has to change — no?

Reply:
I see. Hmm. 2 questions:
1) What does setting canary actually do then? Is that just for debugging output?
2) If we are setting these inline, this is not thread safe unfortunately. Canary is easy to fix by making it atomic, but metadata is not and would need a lock (probably a reader/writer lock if we go that direction).

Thoughts? Whatever we decide we should add more comments here.

Reply:
Also: https://github.com/envoyproxy/envoy/blob/master/source/common/upstream/upstream_impl.h#L69

Reply:
So for:

1) I was setting canary blindly just in case it had changed, given the metadata as a whole change. This is not for debugging, but to address the case in which an endpoint (e.g.: 10.0.0.2:80) had the following transitions in EDS:

```
# eds snapshot v1
10.0.0.2:80 canary:false

# eds snapshot v2
10.0.0.2:80 canary:true
```

That is, it's just a metadata change as everything else and it needs to be update so that subsets can be properly recreated. Was this what you were asking? I suspect maybe not...

2) Ah, I thought this was all computed for each worker and thus thread safe. I'll add locking then. 

And yeah, I'll add more comments too. Lets clear up question 1. before we move forward though.

Reply:
> That is, it's just a metadata change as everything else and it needs to be updated so that subsets can be properly recreated. Was this what you were asking? I suspect maybe not...

In the host description constructor, it makes a new metadata based on the canary field. IIRC canary is only used from that point forward possibly in `/clusters` output but I'm not sure. I would check to see where it's actually used? That's all I was asking. It's not clear to me that updating the canary flag after the metadata was constructed does anything. It makes me wonder if canary should always be pulled from metadata and the getter should be removed also?

Reply:
So it is used in two more places besides the `/clusters` use-case:

`chargeUpstreamCode()`:
https://github.com/envoyproxy/envoy/blob/master/source/common/router/router.cc#L141

`onUpstreamHeaders`:
https://github.com/envoyproxy/envoy/blob/master/source/common/router/router.cc#L625

I would just make the `canary()` getter pull out its value from the metadata on every call, instead of storing it at construction time. Would make my change easier and it also makes it clear that the canary attribute is always bound to the metadata. Thoughts?

Reply:
In those paths there might be some noticeable perf delta for looking in the metadata on every request. For that reason I might leave it and just add some comments as to why.

Reply:
(Though I don't feel that strongly about it either way)

Reply:
@mattklein123 ack — so, one more question: given that metadata must now be lock protected, is it ok to change:

https://github.com/envoyproxy/envoy/blob/master/source/common/upstream/upstream_impl.h#L78

to return a copy of the metadata now? Better option in mind?

Reply:
It's called from here:

https://github.com/envoyproxy/envoy/blob/master/source/common/router/header_formatter.cc#L70

which looks like a perf sensitive path...

Reply:
We could reduce copying the entire metadata to just copying a `ProtobufWkt::Value` by providing:

```
ProtobufWkt::Value metadataValue(...)
```

within HostDescription...

Reply:
I'd have to look in more detail, but I would probably allocate the metadata inside a shared pointer, and then return that. Then, the lock only needs to protect the shared pointer write.

Reply:
Ah yes, that would work.

Reply:
If it's possible we are accessing metadata across threads, do we risk having inconsistent metadata across request processing? E.g. we access field X and Y at different points during LB, we haven't taken a shared pointer snapshot before doing both? Maybe we want the update to apply at request boundary?

Reply:
Hmm, I _think_ the only part in request processing where we access `host->metadata()` is for header formatter:

https://github.com/envoyproxy/envoy/blob/master/source/common/router/header_formatter.cc#L70

But yeah, we could take a snapshot of the shared_ptr as soon as the upstream host is selected — is that what you are thinking? We'd probably have to be careful with retries, etc.

Reply:
@htuch so we could take a snapshot here:

https://github.com/envoyproxy/envoy/blob/master/source/common/request_info/request_info_impl.h#L137

?

Reply:
Should the metadata shared ptr be TLS and we post updates to it, which will then be applied on individual worker threads outside other processing? I think we would still want to latch where you suggest, but we could avoid having to take mutexes and worry about overlap with other execution by doing this.

Reply:
Another alternative might be to treat metadata differences as different hosts -- push a new host with the new metadata into hosts_added and have the old one appear in hosts_removed? I think then we don't have to mess with locking?

Reply:
> Another alternative might be to treat metadata differences as different hosts -- push a new host with the new metadata into hosts_added and have the old one appear in hosts_removed? I think then we don't have to mess with locking?

Yeah, this occurred to me also. If this ends up getting super complicated (which it seems to be) I think we might want to consider this approach. Basically just consider this a new host and do a remove/add operation.

Reply:
Hmm, intuitively I rather not treat this as remove/add... Sounds more heavy-handed than needed.

So do you really think it's too complicated as is now? Shall we explore the remove/add route then?

FWIW, I am testing the current version of this patch now since we really really really need mutable metadata for our setup.

Reply:
I will review your current iteration in a bit. Let's hold here for now until we can review. Thanks for iterating on this. :)